tls/s2n_config.c

```diff
@@ -652,6 +693,135 @@ int s2n_config_get_cert_type(struct s2n_config *config, s2n_cert_type *cert_type
+    }
+
+    if (config->use_tickets == enabled) {
+        return 0;
+    }
+
+    config->use_tickets = enabled;
+
+    if (enabled) {
+        s2n_config_init_session_ticket_keys(config);
+    } else {
+        s2n_config_free_session_ticket_keys(config);
+    }
+
+    return 0;
+}
+
+int s2n_config_set_ticket_encrypt_decrypt_key_lifetime(struct s2n_config *config,
+                                             uint64_t lifetime_in_secs)
+{
+    notnull_check(config);
+
+    config->encrypt_decrypt_key_lifetime_in_nanos = (lifetime_in_secs * ONE_SEC_IN_NANOS);
+    return 0;
+}
+
+int s2n_config_set_ticket_decrypt_key_lifetime(struct s2n_config *config,
+                                                  uint64_t lifetime_in_secs)
+{
+    notnull_check(config);
+
+    config->decrypt_key_lifetime_in_nanos = (lifetime_in_secs * ONE_SEC_IN_NANOS);
+    return 0;
+}
+
+int s2n_config_add_ticket_crypto_key(struct s2n_config *config,
+                                     const uint8_t *name, uint32_t name_len,
+                                     uint8_t *key, uint32_t key_len,
+                                     uint64_t intro_time_in_seconds_from_epoch)
+{
+    notnull_check(config);
+    notnull_check(name);
+    notnull_check(key);
+
+    if (!config->use_tickets) {
+        return 0;
+    }
+
+    GUARD(s2n_config_wipe_expired_ticket_crypto_keys(config, -1));
+
+    S2N_ERROR_IF(key_len == 0, S2N_ERR_INVALID_TICKET_KEY_LENGTH);
+
+    S2N_ERROR_IF(config->ticket_keys->num_of_elements >= S2N_MAX_TICKET_KEYS, S2N_ERR_TICKET_KEY_LIMIT);
+
+    S2N_ERROR_IF(name_len == 0 || name_len > S2N_TICKET_KEY_NAME_LEN || s2n_find_ticket_key(config, name), S2N_ERR_INVALID_TICKET_KEY_NAME_OR_NAME_LENGTH);
+
+    uint64_t now;
+    uint16_t insert_index = 0;
+    uint8_t output_pad[S2N_AES256_KEY_LEN + S2N_TICKET_AAD_IMPLICIT_LEN];
+    struct s2n_blob out_key = { .data = output_pad, .size = sizeof(output_pad) };
+    struct s2n_blob in_key = { .data = key, .size = key_len };
+    struct s2n_blob salt = { .size = 0 };
+    struct s2n_blob info = { .size = 0 };
+
+    struct s2n_blob allocator;
+    struct s2n_ticket_key *session_ticket_key;
+
+    GUARD(s2n_alloc(&allocator, sizeof(struct s2n_ticket_key)));
+    session_ticket_key = (struct s2n_ticket_key *) (void *) allocator.data;
+
+    struct s2n_hmac_state hmac;
+
+    GUARD(s2n_hmac_new(&hmac));
+    GUARD(s2n_hkdf(&hmac, S2N_HMAC_SHA256, &salt, &in_key, &info, &out_key));
+    GUARD(s2n_hmac_free(&hmac));
+
+    struct s2n_hash_state hash = { 0 };
+    uint8_t hash_output[SHA_DIGEST_LENGTH];
+
+    GUARD(s2n_hash_new(&hash));
+    GUARD(s2n_hash_init(&hash, S2N_HASH_SHA1));
+    GUARD(s2n_hash_update(&hash, out_key.data, out_key.size));
+    GUARD(s2n_hash_digest(&hash, hash_output, SHA_DIGEST_LENGTH));
+
+    if (config->ticket_key_hashes->num_of_elements >= S2N_MAX_TICKET_KEY_HASHES) {
+        GUARD(s2n_array_free(config->ticket_key_hashes));
```

Comment:
When array is full, we clean it up and start adding hashes from scratch, so we won't be able to verify uniqueness of keys, right? I'm kind of thinking if we can replace this check with check that new key is not in current key list and save on memory (:

Reply:
> When array is full, we clean it up and start adding hashes from scratch, so we won't be able to verify uniqueness of keys, right?

Right we won't be able to. But, I wonder if this array will get full often. We compare the uniqueness against 10000 key hashes as opposed to 48 keys.

#define S2N_MAX_TICKET_KEYS 48
#define S2N_MAX_TICKET_KEY_HASHES 10000 /* 0.2MB */

Reply:
As discussed, updated 10000 (0.2MB) to 500 (10KB) key hashes. 

We store 10k s2n_configs on a single worker and we have up to 20 nginx workers. So, that is 2GB of dynamically allocated memory. Will revisit this if it's too much.